stan/math/rev/arr/functor/integrate_1d.hpp

```diff
@@ -0,0 +1,143 @@
+#ifndef STAN_MATH_REV_ARR_FUNCTOR_integrate_1d_HPP
+#define STAN_MATH_REV_ARR_FUNCTOR_integrate_1d_HPP
+
+#include <stan/math/prim/scal/err/check_less_or_equal.hpp>
+#include <stan/math/prim/scal/err/domain_error.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/prim/arr/fun/value_of.hpp>
+#include <stan/math/rev/scal/meta/is_var.hpp>
+#include <stan/math/prim/arr/functor/integrate_1d.hpp>
+#include <functional>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Calculate first derivative of f(x, param, std::ostream&)
+ * with respect to the nth parameter. Uses nested reverse mode autodiff
+ */
+template <typename F>
+inline double gradient_of_f(const F& f, const double& x, const double& xc,
+                            const std::vector<double>& theta_vals,
+                            const std::vector<double>& x_r,
+                            const std::vector<int>& x_i, size_t n,
+                            std::ostream& msgs) {
+  double gradient = 0.0;
+  start_nested();
+  std::vector<var> theta_var(theta_vals.size());
+  try {
+    for (size_t i = 0; i < theta_vals.size(); i++)
+      theta_var[i] = theta_vals[i];
+    f(x, xc, theta_var, x_r, x_i, msgs).grad();
+    gradient = theta_var[n].adj();
```

Comment:
We need to have a `if (is_nan(gradient))` here and throw an informative exception if necessary. The question is can we get away with changing gradient from nan to zero if the value of the function is zero?